rust-demangle.c

```diff
@@ -1180,11 +1328,47 @@ bool rust_demangle_with_callback(
     rdm.errored = false;
     rdm.skipping_printing = false;
     rdm.verbose = (flags & RUST_DEMANGLE_FLAG_VERBOSE) != 0;
-    rdm.version = 0;
+    rdm.version = -2; // Invalid version
     rdm.bound_lifetime_depth = 0;
 
+    // Rust symbols always start with R, _R or __R for the v0 scheme or ZN, _ZN
+    // or __ZN for the legacy scheme.
+    if (rdm.sym[0] == '_' && rdm.sym[1] == 'R') {
+        rdm.sym += 2;
+        rdm.version = 0; // v0
+    } else if (rdm.sym[0] == 'R') {
+        // On Windows, dbghelp strips leading underscores, so we accept "R..."
+        // form too.
+        rdm.sym += 1;
+        rdm.version = 0; // v0
+    } else if (rdm.sym[0] == '_' && rdm.sym[1] == '_' && rdm.sym[2] == 'R') {
+        // On OSX, symbols are prefixed with an extra _
+        rdm.sym += 3;
+        rdm.version = 0; // v0
+    } else if (rdm.sym[0] == '_' && rdm.sym[1] == 'Z' && rdm.sym[2] == 'N') {
+        rdm.sym += 3;
+        rdm.version = -1; // legacy
+    } else if (rdm.sym[0] == 'Z' && rdm.sym[1] == 'N') {
+        // On Windows, dbghelp strips leading underscores, so we accept "R..."
+        // form too.
+        rdm.sym += 2;
+        rdm.version = -1; // legacy
+    } else if (rdm.sym[0] == '_' && rdm.sym[1] == '_' && rdm.sym[2] == 'Z' && rdm.sym[3] == 'N') {
+        // On OSX, symbols are prefixed with an extra _
+        rdm.sym += 4;
+        rdm.version = -1; // legacy
+    } else {
+        return false;
+    }
```

Comment:
I think `strncmp` could make some of these more readable (since it roughly allows the equivalent of `.starts_with("...")` IIRC), but it's up to you, this doesn't seem that bad to me.